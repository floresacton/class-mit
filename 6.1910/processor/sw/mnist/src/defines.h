#ifndef DEFINES_H_
#define DEFINES_H_

// Typedefs to make the bit size easier to interpret
typedef signed char int8_t;
typedef unsigned char uint8_t;
typedef signed short int16_t;
typedef unsigned short uint16_t;
typedef signed int int32_t;
typedef unsigned int uint32_t;
typedef long long int64_t;
typedef unsigned long long uint64_t;

#endif