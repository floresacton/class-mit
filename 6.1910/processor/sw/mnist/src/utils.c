#include "utils.h"

void print_string(char* s) {
    while (*s != '\0') {
        print_char(*s);
        s++;
    }
}

const int INT_PRINT_LIMIT = 10;
void print_int(int a) {
    int bases[INT_PRINT_LIMIT];
    bases[0] = 1;
    int i = 0;
    for (i = 1; i < INT_PRINT_LIMIT; i++) {
        bases[i] = bases[i - 1] * 10;
    }
    i = 0;
    if (a == 0) {
        print_char('0');
    } else if (a == 0x80000000) {
        print_string("-2147483648");
    } else {
        if (a < 0) {
            print_char('-');
            a = -a;
        }
        int j = INT_PRINT_LIMIT - 1;
        while (a < bases[j]) j--;
        while (j >= 0) {
            int d = 0;
            while (a >= bases[j]) {
                a -= bases[j];
                d ++;
            }
            print_char('0' + d);
            j--;
        }
    }
}

void print_int_hex(unsigned int a) {
    const char hex_lookup[] = "0123456789abcdef";
    int len = 0;

    if (a == 0) len = 1;
    else for (unsigned int n = a; n; n >>= 4) len++;

    for (int i = len - 1; i >= 0; i--)
        print_char(hex_lookup[(a >> (4 * i)) & 0xf]);
}