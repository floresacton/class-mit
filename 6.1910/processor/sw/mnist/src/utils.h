#ifndef UTILS_H_
#define UTILS_H_


// this is linked  from the crt0.S
void print_char(char x);

void print_string(char* s);

void print_int(int a);

void print_int_hex(unsigned int a);

#endif