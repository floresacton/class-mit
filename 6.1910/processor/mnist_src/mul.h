#ifndef MUL_H_
#define MUL_H_

#include "defines.h"

// GCC uses this unsigned multiplication function to override the * operator for unsigned/signed int multiply
uint32_t __mulsi3 (uint32_t a, uint32_t b) {
    uint32_t result = 0;

    // Examples of how to inline new instructions into C
    // read more at : https://sourceware.org/binutils/docs-2.34/as/RISC_002dV_002dFormats.html
    // and  https://gcc.gnu.org/onlinedocs/gcc/Extended-Asm.html
    //
    // asm volatile (".insn r 0b0110011, 0b000, 0b0000001, %0, %1, %2" : "=r" (result) : "r"(a), "r"(b)); 
 
    while (b) {
        if (b & 1) {
            result += a;
        }
        a = a << 1;
        b = b >> 1;
    }
    return result;

}

// Same as above but for 64 bit multiplication

// WARNING: Using unsigned 32x32 bit multiplication here will cause signed mulitplicaitons
// to fail! Since only 64x64 bit multiplicaiton will work, we suggest you implement and
// call a separate helper function instead which computes signed 32x32 bit multiplicaiton directly.

// If you would like to test your function with the provided testbench,
// make sure to change call to * in sw/mnist/mul64_test.c
uint64_t __muldi3 (uint64_t a, uint64_t b) {
    uint64_t result = 0;
    
    while (b) {
        if (b & 1) {
            result += a;
        }
        a = a << 1;
        b = b >> 1;
    }
    return result;

}

// SUGGESTED: Integrate into model.c and then implement in hardware
int32_t packmul(uint32_t input_i, uint32_t filter_i) {
    // input_i and filter_i each hold four adjacent int8_t numbers cast to a single uint32_t

    // NOTE: Packed elements of input buffer should be treated as unsigned.
    uint8_t a[4]; *(uint32_t*)a = input_i;
    int8_t b[4]; *(uint32_t*)b = filter_i; 

    // NOTE: software implementation of packmul relies on working 32 bit multiplication
    return (
        ((int32_t)a[0] * (int32_t)b[0]) +
        ((int32_t)a[1] * (int32_t)b[1]) +
        ((int32_t)a[2] * (int32_t)b[2]) +
        ((int32_t)a[3] * (int32_t)b[3])
    );
}

#endif