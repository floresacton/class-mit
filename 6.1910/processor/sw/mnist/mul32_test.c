#include "src/defines.h"
#include "src/utils.h"
#include "src/mul.h"

// const volatile prevents constant folding
const volatile int16_t a[] = {-16294,9294,26607,-3540,2308,-21533,-24794,27536,-31106,29908,-8705,-8931,30187,-12890,7320,15385,-15935,-6642,-30858,1700,-11767,-18056,-397,19543,11442,-30011,-29739,29176,-19985,-997,-19011,26892,-29029,-27417,31236,16964,4385,-6461,15966,27924,7752,30512,-3607,-4503,-12599,649,-731,24804,-14526,-9025,6224,-24588,-17664,-27366,-14644,26961,21348,25047,26951,-19061,-3855,-28418,-14301,-17712,4242,8307,-2491,2352,21397,25445,-8046,30672,-16919,-19030,24699,22720,12926,1025,15555,-2817,-22522,-7939,23038,-7426,169,-27510,-6012,23380,-11190,19236,1146,-1641,-21408,-26795,24392,13860,26486,27912,8602,-21305,22112,-24160,31210,-30879,13541,-22953,29453,22623,-31720,-23630,-19382,1454,13846,24323,24892,22910,4341,-14488,11641,2083,-7501,7345,-5151,27710,960,26684,-2100,19593,-8643,27136,32026,9430,21624,23608,-7932,-7849,-24777,592,-5325,15169,-3333,25839,-29370,7209,-28224,-22206,9419,-19550,18576,-598,-5390,11360,11697,-30869,-6517,-29960,-21175,-2163,-29317,-6532,17362,29849,18849,3115,20798,-10985,30462,-7220,-17955,6171,-2230,-24109,9828,26328,30201,-12991,-23225,23647,-7959,-19519,-23387,-16744,17561,6346,-23676,-26660,-11923,-30916,24688,-17828,-19598,-8359,25145,-29843,-10424,5698,22934,7738,31623,-2463};
const volatile int16_t b[] = {22595,20964,20937,-2895,-9975,11374,1116,18392,-22990,23526,-14421,-25766,-23538,-31601,-9973,32734,17662,-32666,29725,10898,-5457,-29352,-13356,30643,-21634,-17864,8716,-23098,12314,-19995,-26572,-26438,4422,22269,-24106,18911,22749,29876,-17050,17098,8293,-6274,31284,-15000,22235,7966,-26460,6054,20313,31065,16897,7253,24440,10159,5198,28707,-13733,11793,-14635,14350,24589,-6987,29216,2732,-16003,-586,-2996,-16316,1021,-10845,11693,27582,-6910,-5641,3334,-16307,-26783,12403,19355,11193,20715,15941,28730,-28126,22723,20697,-16501,10534,12653,2444,-27831,-25192,27804,-25601,2521,-16589,-18588,1594,-10347,-25788,24833,9076,-21531,-6864,-31727,-14061,-5552,-10370,9617,-6040,1311,6573,6732,-27931,-6336,21952,-28665,-21050,26646,-25561,-13030,-5377,2693,12748,-24502,-13269,24784,-24911,13756,27574,13460,-15057,21557,-25172,28808,12027,-2696,11521,4722,21120,5106,32441,26470,12886,-24018,32182,-2935,511,-26887,-13256,4255,-14841,23272,-31573,-31432,-2182,6481,10476,-3195,4191,21850,6623,7229,-6742,29423,-17846,9856,26154,14191,4963,-18930,2281,5028,2989,-15656,3010,-8561,-16451,-17882,-29548,16032,-6188,9839,-28166,-27566,27506,22522,22846,11636,24457,-22870,23571,-1940,-1915,18078,24263,8602,30448,2499,28274};

int main(int argc, char* argv[]) {
    for(int test_no = 0; test_no < sizeof(a)/sizeof(a[0]); test_no ++)  {
        print_string("Input ");
        print_int(test_no);
        print_string(" - ");
        print_int(a[test_no]);
        print_string(" * ");
        print_int(b[test_no]);
        print_string(" = ");
        print_int((int32_t)a[test_no] * (int32_t)b[test_no]);
        print_string("\n");

    }
    return 0;
}
