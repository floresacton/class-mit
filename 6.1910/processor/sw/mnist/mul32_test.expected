Input 0 - -16294 * 22595 = -368162930
Input 1 - 9294 * 20964 = 194839416
Input 2 - 26607 * 20937 = 557070759
Input 3 - -3540 * -2895 = 10248300
Input 4 - 2308 * -9975 = -23022300
Input 5 - -21533 * 11374 = -244916342
Input 6 - -24794 * 1116 = -27670104
Input 7 - 27536 * 18392 = 506442112
Input 8 - -31106 * -22990 = 715126940
Input 9 - 29908 * 23526 = 703615608
Input 10 - -8705 * -14421 = 125534805
Input 11 - -8931 * -25766 = 230116146
Input 12 - 30187 * -23538 = -710541606
Input 13 - -12890 * -31601 = 407336890
Input 14 - 7320 * -9973 = -73002360
Input 15 - 15385 * 32734 = 503612590
Input 16 - -15935 * 17662 = -281443970
Input 17 - -6642 * -32666 = 216967572
Input 18 - -30858 * 29725 = -917254050
Input 19 - 1700 * 10898 = 18526600
Input 20 - -11767 * -5457 = 64212519
Input 21 - -18056 * -29352 = 529979712
Input 22 - -397 * -13356 = 5302332
Input 23 - 19543 * 30643 = 598856149
Input 24 - 11442 * -21634 = -247536228
Input 25 - -30011 * -17864 = 536116504
Input 26 - -29739 * 8716 = -259205124
Input 27 - 29176 * -23098 = -673907248
Input 28 - -19985 * 12314 = -246095290
Input 29 - -997 * -19995 = 19935015
Input 30 - -19011 * -26572 = 505160292
Input 31 - 26892 * -26438 = -710970696
Input 32 - -29029 * 4422 = -128366238
Input 33 - -27417 * 22269 = -610549173
Input 34 - 31236 * -24106 = -752975016
Input 35 - 16964 * 18911 = 320806204
Input 36 - 4385 * 22749 = 99754365
Input 37 - -6461 * 29876 = -193028836
Input 38 - 15966 * -17050 = -272220300
Input 39 - 27924 * 17098 = 477444552
Input 40 - 7752 * 8293 = 64287336
Input 41 - 30512 * -6274 = -191432288
Input 42 - -3607 * 31284 = -112841388
Input 43 - -4503 * -15000 = 67545000
Input 44 - -12599 * 22235 = -280138765
Input 45 - 649 * 7966 = 5169934
Input 46 - -731 * -26460 = 19342260
Input 47 - 24804 * 6054 = 150163416
Input 48 - -14526 * 20313 = -295066638
Input 49 - -9025 * 31065 = -280361625
Input 50 - 6224 * 16897 = 105166928
Input 51 - -24588 * 7253 = -178336764
Input 52 - -17664 * 24440 = -431708160
Input 53 - -27366 * 10159 = -278011194
Input 54 - -14644 * 5198 = -76119512
Input 55 - 26961 * 28707 = 773969427
Input 56 - 21348 * -13733 = -293172084
Input 57 - 25047 * 11793 = 295379271
Input 58 - 26951 * -14635 = -394427885
Input 59 - -19061 * 14350 = -273525350
Input 60 - -3855 * 24589 = -94790595
Input 61 - -28418 * -6987 = 198556566
Input 62 - -14301 * 29216 = -417818016
Input 63 - -17712 * 2732 = -48389184
Input 64 - 4242 * -16003 = -67884726
Input 65 - 8307 * -586 = -4867902
Input 66 - -2491 * -2996 = 7463036
Input 67 - 2352 * -16316 = -38375232
Input 68 - 21397 * 1021 = 21846337
Input 69 - 25445 * -10845 = -275951025
Input 70 - -8046 * 11693 = -94081878
Input 71 - 30672 * 27582 = 845995104
Input 72 - -16919 * -6910 = 116910290
Input 73 - -19030 * -5641 = 107348230
Input 74 - 24699 * 3334 = 82346466
Input 75 - 22720 * -16307 = -370495040
Input 76 - 12926 * -26783 = -346197058
Input 77 - 1025 * 12403 = 12713075
Input 78 - 15555 * 19355 = 301067025
Input 79 - -2817 * 11193 = -31530681
Input 80 - -22522 * 20715 = -466543230
Input 81 - -7939 * 15941 = -126555599
Input 82 - 23038 * 28730 = 661881740
Input 83 - -7426 * -28126 = 208863676
Input 84 - 169 * 22723 = 3840187
Input 85 - -27510 * 20697 = -569374470
Input 86 - -6012 * -16501 = 99204012
Input 87 - 23380 * 10534 = 246284920
Input 88 - -11190 * 12653 = -141587070
Input 89 - 19236 * 2444 = 47012784
Input 90 - 1146 * -27831 = -31894326
Input 91 - -1641 * -25192 = 41340072
Input 92 - -21408 * 27804 = -595228032
Input 93 - -26795 * -25601 = 685978795
Input 94 - 24392 * 2521 = 61492232
Input 95 - 13860 * -16589 = -229923540
Input 96 - 26486 * -18588 = -492321768
Input 97 - 27912 * 1594 = 44491728
Input 98 - 8602 * -10347 = -89004894
Input 99 - -21305 * -25788 = 549413340
Input 100 - 22112 * 24833 = 549107296
Input 101 - -24160 * 9076 = -219276160
Input 102 - 31210 * -21531 = -671982510
Input 103 - -30879 * -6864 = 211953456
Input 104 - 13541 * -31727 = -429615307
Input 105 - -22953 * -14061 = 322742133
Input 106 - 29453 * -5552 = -163523056
Input 107 - 22623 * -10370 = -234600510
Input 108 - -31720 * 9617 = -305051240
Input 109 - -23630 * -6040 = 142725200
Input 110 - -19382 * 1311 = -25409802
Input 111 - 1454 * 6573 = 9557142
Input 112 - 13846 * 6732 = 93211272
Input 113 - 24323 * -27931 = -679365713
Input 114 - 24892 * -6336 = -157715712
Input 115 - 22910 * 21952 = 502920320
Input 116 - 4341 * -28665 = -124434765
Input 117 - -14488 * -21050 = 304972400
Input 118 - 11641 * 26646 = 310186086
Input 119 - 2083 * -25561 = -53243563
Input 120 - -7501 * -13030 = 97738030
Input 121 - 7345 * -5377 = -39494065
Input 122 - -5151 * 2693 = -13871643
Input 123 - 27710 * 12748 = 353247080
Input 124 - 960 * -24502 = -23521920
Input 125 - 26684 * -13269 = -354069996
Input 126 - -2100 * 24784 = -52046400
Input 127 - 19593 * -24911 = -488081223
Input 128 - -8643 * 13756 = -118893108
Input 129 - 27136 * 27574 = 748248064
Input 130 - 32026 * 13460 = 431069960
Input 131 - 9430 * -15057 = -141987510
Input 132 - 21624 * 21557 = 466148568
Input 133 - 23608 * -25172 = -594260576
Input 134 - -7932 * 28808 = -228505056
Input 135 - -7849 * 12027 = -94399923
Input 136 - -24777 * -2696 = 66798792
Input 137 - 592 * 11521 = 6820432
Input 138 - -5325 * 4722 = -25144650
Input 139 - 15169 * 21120 = 320369280
Input 140 - -3333 * 5106 = -17018298
Input 141 - 25839 * 32441 = 838242999
Input 142 - -29370 * 26470 = -777423900
Input 143 - 7209 * 12886 = 92895174
Input 144 - -28224 * -24018 = 677884032
Input 145 - -22206 * 32182 = -714633492
Input 146 - 9419 * -2935 = -27644765
Input 147 - -19550 * 511 = -9990050
Input 148 - 18576 * -26887 = -499452912
Input 149 - -598 * -13256 = 7927088
Input 150 - -5390 * 4255 = -22934450
Input 151 - 11360 * -14841 = -168593760
Input 152 - 11697 * 23272 = 272212584
Input 153 - -30869 * -31573 = 974626937
Input 154 - -6517 * -31432 = 204842344
Input 155 - -29960 * -2182 = 65372720
Input 156 - -21175 * 6481 = -137235175
Input 157 - -2163 * 10476 = -22659588
Input 158 - -29317 * -3195 = 93667815
Input 159 - -6532 * 4191 = -27375612
Input 160 - 17362 * 21850 = 379359700
Input 161 - 29849 * 6623 = 197689927
Input 162 - 18849 * 7229 = 136259421
Input 163 - 3115 * -6742 = -21001330
Input 164 - 20798 * 29423 = 611939554
Input 165 - -10985 * -17846 = 196038310
Input 166 - 30462 * 9856 = 300233472
Input 167 - -7220 * 26154 = -188831880
Input 168 - -17955 * 14191 = -254799405
Input 169 - 6171 * 4963 = 30626673
Input 170 - -2230 * -18930 = 42213900
Input 171 - -24109 * 2281 = -54992629
Input 172 - 9828 * 5028 = 49415184
Input 173 - 26328 * 2989 = 78694392
Input 174 - 30201 * -15656 = -472826856
Input 175 - -12991 * 3010 = -39102910
Input 176 - -23225 * -8561 = 198829225
Input 177 - 23647 * -16451 = -389016797
Input 178 - -7959 * -17882 = 142322838
Input 179 - -19519 * -29548 = 576747412
Input 180 - -23387 * 16032 = -374940384
Input 181 - -16744 * -6188 = 103611872
Input 182 - 17561 * 9839 = 172782679
Input 183 - 6346 * -28166 = -178741436
Input 184 - -23676 * -27566 = 652652616
Input 185 - -26660 * 27506 = -733309960
Input 186 - -11923 * 22522 = -268529806
Input 187 - -30916 * 22846 = -706306936
Input 188 - 24688 * 11636 = 287269568
Input 189 - -17828 * 24457 = -436019396
Input 190 - -19598 * -22870 = 448206260
Input 191 - -8359 * 23571 = -197029989
Input 192 - 25145 * -1940 = -48781300
Input 193 - -29843 * -1915 = 57149345
Input 194 - -10424 * 18078 = -188445072
Input 195 - 5698 * 24263 = 138250574
Input 196 - 22934 * 8602 = 197278268
Input 197 - 7738 * 30448 = 235606624
Input 198 - 31623 * 2499 = 79025877
Input 199 - -2463 * 28274 = -69638862
PASSED